File: scripts/r/causalidad_de_granger_cruzada_multivariable.R
library(tidyverse)
library(car)
library(urca)
library(tseries)
library(lmtest)
library(lubridate)
library(forecast)
library(astsa)
library(vars)
library(mFilter)
library(dynlm)
library(nlme)
library(quantmod)
library(xts)
library(dplyr)

df <- as.data.frame(read_tsv(file = "df.txt"))

variables <- setdiff(names(df), names(df)[1])
frecuencia <- 52
inicio <- c(2008, 1)

series <- lapply(variables, function(v) ts(df[[v]], start = inicio, frequency = frecuencia))
names(series) <- variables

lapply(series, plot)

series_log <- lapply(series, log)

diffs_necesarias <- sapply(series_log, ndiffs)

series_est <- lapply(series_log, function(s) if (ndiffs(s) > 0) diff(s) else s)

sapply(series_est, ndiffs)

ts.plot(do.call(cbind, series_est), col = rainbow(length(series_est)))

adf_results <- as.data.frame(t(sapply(series_est, function(s) adf.test(s)$p.value)))
names(adf_results) <- variables

lag_selection <- lapply(series_est, function(s) VARselect(s, lag.max = 26)$selection)
lag_AIC <- as.data.frame(do.call(cbind, lag_selection))
summary(lag_AIC)

variable_objetivo <- variables[1]
otras_variables <- setdiff(variables, variable_objetivo)
lags <- 1:19

granger_matrix <- expand.grid(objetivo = variable_objetivo, otra = otras_variables, lag = lags, direccion = c("hacia", "desde"))

granger_matrix$p_value <- apply(granger_matrix, 1, function(fila) {
  x <- series_est[[fila[["objetivo"]]]]
  y <- series_est[[fila[["otra"]]]]
  formula_obj <- if (fila[["direccion"]] == "hacia") x ~ y else y ~ x
  grangertest(formula_obj, order = as.numeric(fila[["lag"]]))$`Pr(>F)`[2]
})

write_csv2(granger_matrix, file = "resultado_causalidad_granger.csv")
write_csv2(lag_AIC, file = "resultado_seleccion_lags_aic.csv")
write_csv2(adf_results, file = "resultado_test_adf.csv")

File: scripts/python/causalidad_de_granger_cruzada_multivariable.py
import numpy as np
import pandas as pd
from statsmodels.tsa.stattools import adfuller, grangercausalitytests
from statsmodels.tsa.api import VAR

df = pd.read_csv("df.txt", sep="\t")

variables = [c for c in df.columns if c != df.columns[0]]

series = {v: np.log(df[v]) for v in variables}

for v, s in series.items():
    while adfuller(s.dropna())[1] > 0.05:
        s = s.diff()
    series[v] = s.dropna()

adf_results = {v: adfuller(s)[1] for v, s in series.items()}
adf_results = pd.Series(adf_results)

lag_selection = {}
for v, s in series.items():
    modelo = VAR(pd.DataFrame(s))
    seleccion = modelo.select_order(maxlags=26)
    lag_selection[v] = seleccion.selected_orders

lag_AIC = pd.DataFrame(lag_selection)

variable_objetivo = variables[0]
otras_variables = [v for v in variables if v != variable_objetivo]
lags = range(1, 20)

registros = []
for otra in otras_variables:
    largo = min(len(series[variable_objetivo]), len(series[otra]))
    datos = pd.concat([series[variable_objetivo].iloc[-largo:], series[otra].iloc[-largo:]], axis=1)
    datos.columns = [variable_objetivo, otra]

    for lag in lags:
        for direccion, orden_columnas in [("hacia", [variable_objetivo, otra]), ("desde", [otra, variable_objetivo])]:
            resultado = grangercausalitytests(datos[orden_columnas], maxlag=lag, verbose=False)
            p_value = resultado[lag][0]["ssr_ftest"][1]
            registros.append({"objetivo": variable_objetivo, "otra": otra, "lag": lag, "direccion": direccion, "p_value": p_value})

granger_matrix = pd.DataFrame(registros)

granger_matrix.to_csv("resultado_causalidad_granger.csv", sep=";", index=False)
lag_AIC.to_csv("resultado_seleccion_lags_aic.csv", sep=";", index=False)
adf_results.to_csv("resultado_test_adf.csv", sep=";")
